### PR TITLE
[InstCombine] Simplify `(X / C0) * C1 + (X % C0) * C2` to `(X / C0) * (C1 - C2 * C0) + X * C2`

### LINKED ISSUE #76128: [Reassociate] Missing optimization: fold `div(v, a) * b + rem(v, a)` to `div(v, a) * (b - a) + v`
Alive2 proof: https://alive2.llvm.org/ce/z/7SnGmc

### Description:

```llvm
define i32 @src(i32 noundef %val) {
entry:
  %div = udiv i32 %val, 10
  %shl = shl i32 %div, 4
  %rem = urem i32 %val, 10
  %add = or disjoint i32 %shl, %rem
  ret i32 %add
}
```
can be folded to:
```llvm
define i32 @tgt(i32 noundef %val) {
entry:
  %div = udiv i32 %val, 10
  %reass.mul = mul nuw i32 %div, 6
  %i = add i32 %reass.mul, %val
  ret i32 %i
}
```
This is a simple example. `or disjoint` can be replaced with `add nsw nuw`, `shl 4` can be replaced with equivalent `mul 16`, and constants here can be replaced with other simple constants: https://alive2.llvm.org/ce/z/qmYLyJ
I don't replace constants with arguments because I haven't figure out completetly what condition is needed for this fold yet.

### Real-world motivation

This snippet of IR is derived from [linux/lib/bcd.c](https://github.com/torvalds/linux/blob/master/lib/bcd.c#L11C17-L11C17) (after O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, see also:https://godbolt.org/z/hra9qYbY7

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by dtcxzyw**
Since `DivRemPairPass` runs after `ReassociatePass` in the optimization pipeline, I decided to do this simplification in `InstCombine`.

Alive2: https://alive2.llvm.org/ce/z/Jgsiqf
Fixes #76128.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 1]
```diff
@@ -3872,6 +3872,10 @@ Instruction *InstCombinerImpl::visitOr(BinaryOperator &I) {
     }
   }
 
+  if (cast<PossiblyDisjointInst>(I).isDisjoint())
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3875]: It is ugly. But I don't know the best approach to treat `or disjoint` as `add nsw nuw`.

It is a phase ordering issue. We cannot assume that `ReassociatePass` will revert the canonicalization `add -> or disjoint`. BTW, I see this change achieves higher coverage in my benchmark (in `openexr/ImfTimeCode.cpp` and `z3/mpf.cpp`).

**Comment by dtcxzyw** [comment]
Ping.

**Comment by nikic** [comment]
As it took me a while to get it: The core of the transform is that `x % c` is `x - x / c * c` (https://alive2.llvm.org/ce/z/I8CBWy).

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3914]: We should probably have something like visitAddLike that is called from both visitAdd and visitOr where we can place folds we want to apply both to add and or disjoint?

**Review by nikic — COMMENTED** [review]
This looks fine to me but needs a rebase.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3961]: a less ugly way at least would be use `m_AddLike` matcher. But I agree with nikic.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 2]
```diff
@@ -1161,6 +1163,33 @@ Value *InstCombinerImpl::SimplifyAddWithRemainder(BinaryOperator &I) {
...
+    Rem = RHS, C2 = APInt(I.getType()->getScalarSizeInBits(), 1);
+  if (match(Div, m_IRem(m_Value(), m_Value()))) {
+    std::swap(Div, Rem);
+    std::swap(C1, C2);
+  }
+  Value *DivOpV;
+  APInt DivOpC;
+  if (MatchRem(Rem, X, C0, IsSigned) &&
+      MatchDiv(Div, DivOpV, DivOpC, IsSigned) && X == DivOpV && C0 == DivOpC) {
+    if (!isGuaranteedNotToBeUndef(X, &AC, &I, &DT))
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1181]: I imagine this isn't super hot code, but the no-undef check should be after the constant checks as its by far the most expensive.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 2 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp]: Done.

**Pull request merged**